prefect_azure/blob_storage.py

```diff
@@ -177,3 +191,517 @@ def example_blob_storage_list_flow():
+        return (Path(self.base_folder) / Path(path)).as_posix()
+
+    @sync_compatible
+    async def download_folder_to_path(
+        self,
+        from_folder: str,
+        to_folder: Union[str, Path],
+        **download_kwargs: Dict[str, Any],
+    ) -> Coroutine[Any, Any, Path]:
+        """Download a folder from the container to a local path.
+
+        Args:
+            from_folder: The folder path in the container to download.
+            to_folder: The local path to download the folder to.
+            **download_kwargs: Additional keyword arguments passed into
+                `BlobClient.download_blob`.
+
+        Returns:
+            The local path where the folder was downloaded.
+
+        Example:
+            Download the contents of container folder `folder` from the container
+                to the local folder `local_folder`:
+
+            ```python
+            from prefect_azure import AzureBlobStorageCredentials
+            from prefect_azure.blob_storage import AzureBlobStorageContainer
+
+            credentials = AzureBlobStorageCredentials(
+                connection_string="connection_string",
+            )
+            block = AzureBlobStorageContainer(
+                container_name="container",
+                credentials=credentials,
+            )
+            block.download_folder_to_path(
+                from_folder="folder",
+                to_folder="local_folder"
+            )
+            ```
+        """
+        self.logger.info(
+            "Downloading folder from container %s to path %s",
+            self.container_name,
+            to_folder,
+        )
+        full_container_path = self._get_path_relative_to_base_folder(from_folder)
+        async with self.credentials.get_container_client(
+            self.container_name
+        ) as container_client:
+            try:
+                async for blob in container_client.list_blobs(
+                    name_starts_with=full_container_path
+                ):
+                    blob_path = blob.name
+                    local_path = Path(to_folder) / Path(blob_path).relative_to(
+                        full_container_path
+                    )
+                    local_path.parent.mkdir(parents=True, exist_ok=True)
+                    async with container_client.get_blob_client(
+                        blob_path
+                    ) as blob_client:
+                        blob_obj = await blob_client.download_blob(**download_kwargs)
+
+                    with local_path.open(mode="wb") as to_file:
+                        await blob_obj.readinto(to_file)
+            except ResourceNotFoundError as exc:
+                raise RuntimeError(
+                    "An error occurred when attempting to download from container"
+                    f" {self.container_name}: {exc.reason}"
+                ) from exc
+
+        return Path(to_folder)
+
+    @sync_compatible
+    async def download_object_to_file_object(
```

Comment:
Same not super cute nonexistent `from_path` error but certainly not a big deal:
```bash
❯ python qa.py
13:20:36.421 | INFO    | prefect.AzureBlobStorageContainer - Downloading object from container testing314 to file object
Traceback (most recent call last):
  File "/Users/bean/code-oss/prefect-azure/qa.py", line 14, in <module>
    container.download_object_to_file_object(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/utilities/asyncutils.py", line 259, in coroutine_wrapper
    return call()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 431, in __call__
    return self.result()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 317, in result
    return self.future.result(timeout=timeout)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 178, in result
    return self.__get_result()
  File "/Users/bean/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 388, in _run_async
    result = await coro
  File "/Users/bean/code-oss/prefect-azure/prefect_azure/blob_storage.py", line 348, in download_object_to_file_object
    blob_obj = await blob_client.download_blob(**download_kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/tracing/decorator_async.py", line 77, in wrapper_use_tracer
    return await func(*args, **kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_blob_client_async.py", line 544, in download_blob
    await downloader._setup()  # pylint: disable=protected-access
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_download_async.py", line 289, in _setup
    self._response = await self._initial_request()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_download_async.py", line 384, in _initial_request
    process_storage_error(error)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_shared/response_handlers.py", line 184, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
  File "<string>", line 1, in <module>
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_download_async.py", line 338, in _initial_request
    location_mode, response = await self._clients.blob.download(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/tracing/decorator_async.py", line 77, in wrapper_use_tracer
    return await func(*args, **kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_generated/aio/operations/_blob_operations.py", line 210, in download
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ResourceNotFoundError: The specified blob does not exist.
RequestId:3b6a9eb5-501e-005d-4e3c-768d9a000000
Time:2024-03-14T18:20:36.6499795Z
ErrorCode:BlobNotFound
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>BlobNotFound</Code><Message>The specified blob does not exist.
RequestId:3b6a9eb5-501e-005d-4e3c-768d9a000000
Time:2024-03-14T18:20:36.6499795Z</Message></Error>
```

Reply:
Error message should be improved by https://github.com/PrefectHQ/prefect-azure/pull/139/commits/99da8e26b622c4d3dcc08a1b2a4dd1bf781ee8de

Reply:
yay!
```bash
❯ python qa.py
13:32:48.711 | INFO    | prefect.AzureBlobStorageContainer - Downloading object from container testing314 to file object
Traceback (most recent call last):
  File "/Users/bean/code-oss/prefect-azure/prefect_azure/blob_storage.py", line 360, in download_object_to_file_object
    blob_obj = await blob_client.download_blob(**download_kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/tracing/decorator_async.py", line 77, in wrapper_use_tracer
    return await func(*args, **kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_blob_client_async.py", line 544, in download_blob
    await downloader._setup()  # pylint: disable=protected-access
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_download_async.py", line 289, in _setup
    self._response = await self._initial_request()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_download_async.py", line 384, in _initial_request
    process_storage_error(error)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_shared/response_handlers.py", line 184, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
  File "<string>", line 1, in <module>
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_download_async.py", line 338, in _initial_request
    location_mode, response = await self._clients.blob.download(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/tracing/decorator_async.py", line 77, in wrapper_use_tracer
    return await func(*args, **kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_generated/aio/operations/_blob_operations.py", line 210, in download
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ResourceNotFoundError: The specified blob does not exist.
RequestId:8505cba1-601e-0069-193e-76be52000000
Time:2024-03-14T18:32:48.9440092Z
ErrorCode:BlobNotFound
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>BlobNotFound</Code><Message>The specified blob does not exist.
RequestId:8505cba1-601e-0069-193e-76be52000000
Time:2024-03-14T18:32:48.9440092Z</Message></Error>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bean/code-oss/prefect-azure/qa.py", line 15, in <module>
    container.download_object_to_file_object(from_path="test.txt", to_file_object="test.txt")
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/utilities/asyncutils.py", line 259, in coroutine_wrapper
    return call()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 431, in __call__
    return self.result()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 317, in result
    return self.future.result(timeout=timeout)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 178, in result
    return self.__get_result()
  File "/Users/bean/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 388, in _run_async
    result = await coro
  File "/Users/bean/code-oss/prefect-azure/prefect_azure/blob_storage.py", line 363, in download_object_to_file_object
    raise RuntimeError(
RuntimeError: An error occurred when attempting to download from container testing314: The specified blob does not exist.
```